isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java

```diff
@@ -89,22 +109,128 @@ public RelNode visit(NamedScan namedScan) throws RuntimeException {
   public RelNode visit(Project project) throws RuntimeException {
     RelNode child = project.getInput().accept(this);
     List<RexNode> rexList =
-        project.getExpressions().stream()
+        project.getExpressions().stream().map(expr -> expr.accept(expressionRexConverter)).toList();
+
+    return relBuilder.push(child).project(rexList).build();
+  }
+
+  @Override
+  public RelNode visit(Join join) throws RuntimeException {
+    var left = join.getLeft().accept(this);
+    var right = join.getRight().accept(this);
+    var condition =
+        join.getCondition()
+            .map(c -> c.accept(expressionRexConverter))
+            .orElse(relBuilder.literal(true));
+    var joinType =
+        switch (join.getJoinType()) {
+          case INNER -> JoinRelType.INNER;
+          case LEFT -> JoinRelType.LEFT;
+          case RIGHT -> JoinRelType.RIGHT;
+          case OUTER -> JoinRelType.FULL;
+          case SEMI -> JoinRelType.SEMI;
+          case ANTI -> JoinRelType.ANTI;
+          case UNKNOWN -> throw new UnsupportedOperationException(
+              "Unknown join type is not supported");
+        };
+    return relBuilder.push(left).push(right).join(joinType, condition).build();
+  }
+
+  @Override
+  public RelNode visit(Aggregate aggregate) throws RuntimeException {
+    RelNode child = aggregate.getInput().accept(this);
+    var groupExprLists =
+        aggregate.getGroupings().stream()
             .map(
-                expr ->
-                    expr.accept(
-                        new ExpressionRexConverter(
-                            relOptCluster.getTypeFactory(), scalarFunctionConverter)))
+                gr ->
+                    gr.getExpressions().stream()
+                        .map(expr -> expr.accept(expressionRexConverter))
+                        .toList())
             .toList();
+    List<RexNode> groupExprs =
```

Comment:
It feels like the grouping set dynamics here are wrong. Do you have a test case to verify the correct conversion?

Reply:
Also, how do we map the dynamics of grouping set identifiers between the Calcite and Substrait? (In Substrait we identify the grouping set the result is from by referencing it via ordinal).

Reply:
`Substrait2SqlTest::simpleTestGroupingSets` has tests covering query with grouping sets.  including `()`. It also has query with `grouping sets` + `group by`.

Substrait uses ordinal, while Calcite uses bitsets.  When converting from substrait to calcite, ordinals are converted into RexInputReference.  The list of RexInputReference in `groupExprs` will converted into indexes in `bitsets`, when we call:

 `RelBuilder.groupKey(Iterable<? extends RexNode> nodes,
      Iterable<? extends Iterable<? extends RexNode>> nodeLists)`

Reply:
I just added one group sets test, same column appear in the group sets multiple times. It seems the conversion work correctly.